exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Otel2PrometheusConverter.java

```diff
@@ -0,0 +1,547 @@
+    return Quantiles.of(result);
+  }
+
+  @Nullable
+  private Exemplar convertLongExemplar(List<LongExemplarData> exemplars) {
+    if (exemplars.size() == 0) {
+      return null;
+    } else {
+      LongExemplarData exemplar = exemplars.get(0);
+      return convertExemplar((double) exemplar.getValue(), exemplar);
+    }
+  }
+
+  @Nullable
+  private Exemplar convertDoubleExemplar(List<DoubleExemplarData> exemplars) {
+    if (exemplars.size() == 0) {
+      return null;
+    } else {
+      DoubleExemplarData exemplar = exemplars.get(0);
+      return convertExemplar(exemplar.getValue(), exemplar);
+    }
+  }
+
+  private Exemplars convertDoubleExemplars(List<DoubleExemplarData> exemplars) {
+    Exemplar[] result = new Exemplar[exemplars.size()];
+    for (int i = 0; i < result.length; i++) {
+      DoubleExemplarData exemplar = exemplars.get(i);
+      result[i] = convertExemplar(exemplar.getValue(), exemplar);
+    }
+    return Exemplars.of(result);
+  }
+
+  private Exemplar convertExemplar(double value, ExemplarData exemplar) {
+    SpanContext spanContext = exemplar.getSpanContext();
+    if (spanContext.isValid()) {
+      return new Exemplar(
+          value,
+          convertAttributes(
+              null,
+              exemplar.getFilteredAttributes(),
+              "trace_id",
+              spanContext.getTraceId(),
+              "span_id",
+              spanContext.getSpanId()),
+          exemplar.getEpochNanos() / NANOS_PER_MILLISECOND);
+    } else {
+      return new Exemplar(
+          value,
+          convertAttributes(null, exemplar.getFilteredAttributes()),
+          exemplar.getEpochNanos() / NANOS_PER_MILLISECOND);
+    }
+  }
+
+  private InfoSnapshot makeTargetInfo(Resource resource) {
+    return new InfoSnapshot(
+        new MetricMetadata("target"),
+        Collections.singletonList(
+            new InfoDataPointSnapshot(convertAttributes(null, resource.getAttributes()))));
+  }
+
+  private InfoSnapshot makeScopeInfo(Set<InstrumentationScopeInfo> scopes) {
+    List<InfoDataPointSnapshot> prometheusScopeInfos = new ArrayList<>(scopes.size());
+    for (InstrumentationScopeInfo scope : scopes) {
+      prometheusScopeInfos.add(
+          new InfoDataPointSnapshot(convertAttributes(scope, scope.getAttributes())));
+    }
+    return new InfoSnapshot(new MetricMetadata("otel_scope"), prometheusScopeInfos);
+  }
+
+  private Labels convertAttributes(
+      @Nullable InstrumentationScopeInfo scope,
+      Attributes attributes,
+      String... additionalAttributes) {
+    int numberOfScopeAttributes = 0;
+    if (otelScopeEnabled && scope != null) {
+      numberOfScopeAttributes = scope.getVersion() == null ? 1 : 2;
+    }
+    String[] names =
+        new String[attributes.size() + numberOfScopeAttributes + additionalAttributes.length / 2];
+    String[] values = new String[names.length];
+    int[] pos = new int[] {0}; // using an array because we want to increment in a forEach() lambda.
```

Comment:
Not sure. Maybe a performance specialist could create an example where `AtomicInteger` is slower in multi threaded scenarios because it syncs with main memory? Anyway, in this example I don't think it makes a difference.